storage/node_liveness_test.go

```diff
@@ -0,0 +1,172 @@
+	})
+}
+
+func TestNodeLiveness(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	mtc := startMultiTestContext(t, 3)
+	defer mtc.Stop()
+
+	// Verify liveness of all nodes for all nodes.
+	verifyLiveness(t, mtc)
+
+	// Advance clock past the liveness threshold to verify IsLive becomes false.
+	mtc.manualClock.Increment(storage.LivenessThreshold.Nanoseconds() + 1)
+	for idx, nl := range mtc.nodeLivenesses {
+		nodeID := mtc.gossips[idx].GetNodeID()
+		live, err := nl.IsLive(nodeID)
+		if live {
+			t.Errorf("expected node %d to be considered not-live after advancing node clock", nodeID)
+		} else if err != nil {
+			t.Error(err)
+		}
+	}
+	// Trigger a manual heartbeat and verify liveness is reestablished.
+	for _, nl := range mtc.nodeLivenesses {
+		nl.ManualHeartbeat()
+	}
+	verifyLiveness(t, mtc)
+}
+
+// TestNodeLivenessIncrement verifies that incrementing the epoch of a
+// node requires the node to be considered not-live and that on
+// increment, no other nodes believe the epoch-incremented node to be
+// live (that is, its received timestamp is not advanced.
+func TestNodeLivenessIncrement(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	mtc := startMultiTestContext(t, 2)
+	defer mtc.Stop()
+	verifyLiveness(t, mtc)
+
+	// First try to increment the epoch of a known-live node.
+	deadNodeID := mtc.gossips[1].GetNodeID()
+	if err := mtc.nodeLivenesses[0].IncrementEpoch(context.Background(), deadNodeID); err == nil {
+		t.Fatalf("expected error incrementing a live node")
+	}
+
+	// Advance clock past liveness threshold & increment epoch.
+	oldLiveness, err := mtc.nodeLivenesses[0].GetLiveness(deadNodeID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	mtc.manualClock.Increment(storage.LivenessThreshold.Nanoseconds() + 1)
+	if err := mtc.nodeLivenesses[0].IncrementEpoch(context.Background(), deadNodeID); err != nil {
+		t.Fatalf("unexpected error incrementing a live node: %s", err)
+	}
+
+	// Verify that the epoch has been advanced
+	util.SucceedsSoon(t, func() error {
+		newLiveness, err := mtc.nodeLivenesses[0].GetLiveness(deadNodeID)
+		if err != nil {
+			return err
+		}
+		if newLiveness.Epoch != oldLiveness.Epoch+1 {
+			return errors.Errorf("expected epoch to increment")
+		}
+		if newLiveness.LastHeartbeat != oldLiveness.LastHeartbeat {
+			return errors.Errorf("expected last heartbeat to remain unchanged")
+		}
+		if live, err := mtc.nodeLivenesses[0].IsLive(deadNodeID); live || err != nil {
+			return errors.Errorf("expected dead node to remain dead after epoch increment: %s", err)
+		}
+		return nil
+	})
+}
+
+// TestNodeLivenessRestart verifies that if nodes are shutdown and
+// restarted, the node liveness records are re-gossiped immediately.
+func TestNodeLivenessRestart(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	mtc := startMultiTestContext(t, 1)
+	defer mtc.Stop()
+
+	// After verifying node is in liveness table, stop store.
+	verifyLiveness(t, mtc)
+	mtc.stopStore(0)
+
+	// Create a new gossip instance and connect it.
```

Comment:
Could you just make two stores in the multiTestContext instead of creating a second gossip instance by hand?